mountpoint-s3/src/fs.rs

```diff
@@ -425,30 +428,92 @@ impl S3Personality {
+            sse_kms_key_id: Default::default(),
+            checksum: Crc32c::new(0),
+        }
+    }
 }
 
 impl ServerSideEncryption {
     /// Construct SSE settings from raw values provided via CLI
     pub fn new(sse_type: Option<String>, sse_kms_key_id: Option<String>) -> Self {
-        // TODO: compute checksum
+        let checksum = Self::compute_checksum(sse_type.as_deref(), sse_kms_key_id.as_deref());
         Self {
             sse_type,
             sse_kms_key_id,
+            checksum,
         }
     }
 
-    /// String representation of the SSE type as it is expected by S3 API
-    pub fn sse_type(&self) -> Option<String> {
-        self.sse_type.clone()
+    /// Computes the checksum of SSE settings by combining two strings containing the type and the key
+    /// Note, that this implementation yields the same result for Some("") and None, but we may safely
+    /// assume that it will never be called with an empty string as one of its parameters.
+    fn compute_checksum(sse_type: Option<&str>, sse_kms_key_id: Option<&str>) -> Crc32c {
+        let mut hasher: Hasher = Hasher::new();
+        if let Some(maybe_sse_type) = sse_type {
+            hasher.update(maybe_sse_type.as_bytes().as_ref());
+        }
+        if let Some(maybe_sse_kms_key_id) = sse_kms_key_id {
+            hasher.update(maybe_sse_kms_key_id.as_bytes().as_ref());
+        }
+        hasher.finalize()
     }
 
-    /// AWS KMS Key ID, if provided
-    pub fn key_id(&self) -> Option<String> {
-        self.sse_kms_key_id.clone()
+    /// Checks that SSE settings still match the checksum and returns the string representations of:
+    /// 1. the SSE type as it is expected by S3 API and;
+    /// 2. AWS KMS Key ID, if provided.
+    pub fn into_inner(self) -> Result<(Option<String>, Option<String>), IntegrityError> {
+        let computed = Self::compute_checksum(self.sse_type.as_deref(), self.sse_kms_key_id.as_deref());
+        if computed == self.checksum {
+            Ok((self.sse_type, self.sse_kms_key_id))
+        } else {
+            Err(IntegrityError::ChecksumMismatch(self.checksum, computed))
+        }
+    }
+
+    /// Checks that values provided as arguments to this function match the checksum stored in the object.
+    /// S3 will return some values for sse type and key even if they were not set on our side. We want to check
+    /// only the values which we set, thus response values are only added to checksum if they are present in this object.
+    pub fn verify_response(&self, sse_type: Option<&str>, sse_kms_key_id: Option<&str>) -> Result<(), IntegrityError> {
+        // we first validate in-memory values, since we are using them to compute a response's checksum
+        let computed = Self::compute_checksum(self.sse_type.as_deref(), self.sse_kms_key_id.as_deref());
+        if computed != self.checksum {
+            return Err(IntegrityError::ChecksumMismatch(self.checksum, computed));
+        }
```

Comment:
nit: we could extract this into a `validate` methods (used here and in `into_inner`).